Mathlib/Computability/ContextFreeGrammar.lean

```diff
@@ -187,3 +190,76 @@ lemma Derives.append_right {v w : List (Symbol T g.NT)}
   | tail _ last ih => exact ih.trans_produces <| last.append_right p
 
 end ContextFreeGrammar
+
+/-- Context-free languages are defined by context-free grammars. -/
+def Language.IsContextFree (L : Language T) : Prop :=
+  ∃ g : ContextFreeGrammar.{uT} T, g.language = L
+
+section closure_reversal
+
+private def reverseRule {N : Type uN} (r : ContextFreeRule T N) : ContextFreeRule T N :=
```

Comment:
Why private (same below)? IMHO, a public `ContextFreeGrammar.reverse` with a theorem that it gives the reverse language is stronger than the existence theorem.
```suggestion
protected def ContextFreeRule.reverse {N : Type uN} (r : ContextFreeRule T N) : ContextFreeRule T N :=
```

Reply:
I don't want to expose implementation. What I want to export out of this file is that "reversal can be done".

Reply:
Why? People may want to prove other theorems about this operation and its interaction with other operations on context free grammars.

Reply:
Then let's wait for someone who uses this part of the library... My impression about `Mathlib` is that the only reason for a lemma to be private is that nobody will ever need it (e.g., because it is an auxiliary lemma for a stronger statement). Since we have no backwards compatibility promises (yet), it doesn't even force us to maintain this API in the future (unless someone will actually use it).

BTW, the private version applies to all universes while the public version only works for the same universe.

Reply:
How it is now.

Reply:
What would I have to do for everything to apply to all universes?

Reply:
You have a predicate that says that there exists a context-free grammar in the same universe that defines this language. You can turn it into a universe-polymorphic predicate (and possibly prove some theorems using `UnivLE`) or make the API behind the proof public (i.e., add public definition/lemmas `ContextFreeRule.reverse`, `ContextFreeGrammar.reverse`, `Derives.reverse`, `ContextFreeGrammar.language_reverse`) and let the (possibly, empty set of) users who care about universe-polymorphism here handle it as they want.

Reply:
I refactored it in such a way that everything can be public now.